zebra-consensus/src/verify/redjubjub.rs

```diff
@@ -0,0 +1,149 @@
+        // XXX(hdevalence) what's a reasonable choice here?
+        let (tx, _) = channel(10);
+        Self { tx, batch }
+    }
+}
+
+/// Type alias to clarify that this batch::Item is a RedJubjubItem
+pub type RedJubjubItem<T> = batch::Item<T>;
+
+impl<'msg, T: SigType> Service<BatchControl<RedJubjubItem<T>>> for RedJubjubVerifier<T> {
+    type Response = ();
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: BatchControl<RedJubjubItem<T>>) -> Self::Future {
+        match req {
+            BatchControl::Item(item) => {
+                tracing::trace!("got item");
+                self.batch.queue(item);
+                let mut rx = self.tx.subscribe();
+                Box::pin(async move {
+                    match rx.recv().await {
+                        Ok(result) => result,
+                        Err(RecvError::Lagged(_)) => {
+                            tracing::warn!(
+                                "missed channel updates for the correct signature batch!"
+                            );
+                            Err(Error::InvalidSignature)
+                        }
+                        Err(RecvError::Closed) => panic!("verifier was dropped without flushing"),
+                    }
+                })
+            }
+            BatchControl::Flush => {
+                tracing::trace!("got flush command");
+                let batch = mem::take(&mut self.batch);
+                let _ = self.tx.send(batch.verify(thread_rng()));
+                Box::pin(async { Ok(()) })
+            }
+        }
+    }
+}
+
+impl<T: SigType> Drop for RedJubjubVerifier<T> {
+    fn drop(&mut self) {
+        // We need to flush the current batch in case there are still any pending futures.
+        let batch = mem::take(&mut self.batch);
+        let _ = self.tx.send(batch.verify(thread_rng()));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::time::Duration;
+
+    use color_eyre::eyre::Result;
+    use futures::stream::{FuturesUnordered, StreamExt};
+    use tower::ServiceExt;
+    use tower_batch::Batch;
+
+    async fn sign_and_verify<V, T>(mut verifier: V, n: usize) -> Result<(), V::Error>
+    where
+        T: SigType,
+        V: Service<RedJubjubItem<T>, Response = ()>,
+    {
+        let rng = thread_rng();
+        let mut results = FuturesUnordered::new();
+        for i in 0..n {
+            let span = tracing::trace_span!("sig", i);
+            let sk = SigningKey::<T>::new(rng);
+            let vk = VerificationKey::from(&sk);
+            let msg = b"BatchVerifyTest";
+            let sig = sk.sign(rng, &msg[..]);
+
+            verifier.ready_and().await?;
+            results.push(span.in_scope(|| verifier.call((vk.into(), sig, msg).into())))
```

Comment:
I don't think this `in_scope` does what is intended, this would enter `span` for the duration of call, which then returns a future, but then it would push an uninstrumented future into the `results`. If you want the `span` to be active while the future is being polled you should use `instrument`
```suggestion
            results.push(verifier.call((vk.into(), sig, msg).into()).instrument(span))
```